bindings/cs/rl.net/Native/SenderAdapter.cs

```diff
@@ -0,0 +1,145 @@
+
+        public SenderAdapter(ISender senderImplementation)
+        {
+            this.senderImplementation = senderImplementation;
+        }
+
+        public static sender_vtable VTable
+        {
+            get;
+        } = CreateAdapterVTable();
+
+        private static sender_vtable CreateAdapterVTable()
+        {
+            return new sender_vtable
+            {
+                init = (managed_handle, status) => InvokeInit(managed_handle, status),
+                send = (managed_handle, buffer, status) => InvokeSend(managed_handle, buffer, status),
+                release = (managed_handle) => InvokeRelease(managed_handle)
+            };
+        }
+
+        private static SenderAdapter GetAdapterOrThrow(IntPtr managed_handle)
+        {
+            GCHandle gcHandle = GCHandle.FromIntPtr(managed_handle);
+            SenderAdapter adapter = gcHandle.Target as SenderAdapter;
+
+            // This should never happen, unless there is a bug in the binding implementation.
+            if (adapter == null)
+            {
+                StringBuilder errorBuilder = new StringBuilder(NativeMethods.OpaqueBindingErrorMessage)
+                    .Append("Invalid managed_handle in ISender binding adapter. Expecting non-null RL.Net.Native.SenderAdapter. Got ")
+                    .AppendFormat("{0}.", gcHandle.Target == null ? "null" : gcHandle.Target.GetType().FullName);
+
+                throw new RLException(errorBuilder.ToString());
+            }
+
+            return adapter;
+        }
+
+        private static int InvokeAndUnwrapExceptions(Func<ApiStatus, int> fn, IntPtr status)
+        {
+            using (ApiStatus apiStatus = new ApiStatus(status))
+            {
+                try
+                {
+                    return fn(apiStatus);
+                }
+                catch (RLException e)
+                {
+                    return e.UpdateApiStatus(apiStatus);
+                }
+                catch (Exception e)
+                {
+                    return new ApiStatusBuilder(NativeMethods.OpaqueBindingError)
+                        .AppendLine(e.Message)
+                        .AppendLine(e.StackTrace)
+                        .UpdateApiStatus(apiStatus);
+                }
+            }
+        }
+
+        private static int InvokeInit(IntPtr managed_handle, IntPtr status)
+        {
+            return SenderAdapter.InvokeAndUnwrapExceptions(
+                (ApiStatus apiStatus) =>
+                {
+                    SenderAdapter adapter = GetAdapterOrThrow(managed_handle);
+
+                    adapter.senderImplementation.Init(apiStatus);
+                    return apiStatus.ErrorCode;
+                },
+                status
+            );
+        }
+
+        private static int InvokeSend(IntPtr managed_handle, IntPtr buffer, IntPtr status)
+        {
+            return SenderAdapter.InvokeAndUnwrapExceptions(
+                (ApiStatus apiStatus) =>
```

Comment:
The `ApiStatus` argument is not really necessary since we're capturing `buffer` and `managed_handle` anyways

Reply:
It seems a bit cleaner to have `InvokeAndUnwrapExceptions` be responsible for creating the `ApiStatus` managed wrapper and managing its lifetime - otherwise it would just be duplicated code across `InvokeInit` and `InvokeSend`